Refactor Account Settings Screen - Creates AccountSettingsViewModel

SowmiaS opened the pull request:
Part of #12372

### **Description**

The Second PR for the task of refactoring the `AccountSettingsFragment` to MVVM pattern following the reactive approach using Flow. This PR creates the ViewModel for the AccountSettingsScreen.

- `AccountSettingsViewModel` maintains the `AccountSettingsUiState`.  The state is updated with the account information
from DB after applying optimistically changed preferences.
-  Receives the UI events from `AccountSettingsFragment`. It processes the UI events and updates the `AccountSettingsUiState` with the updated account information. 
- The UI Events are 
   - onUsernameChangeConfirmedFromServer
   - onPrimarySiteChanged
   - onEmailChanged
   - onWebAddressChanged
   - onPasswordChanged
- On initialization, the AccountSettingsViewModel requests to fetch new account settings and gets all sites from DB.

Note: As it is the settings page, the user is given the impression that the settings are changed immediately by updating the UI state before updating the server. In case of any error, it is reverted back to the UI state in alignment with the Account information available from the server.  


This PR also has a small change related to PR #16368. This PR moves the dispatcher.unregister() to `onCleared` method.

### **Testing**

- No testing is required.
- Unit test is written to test the code.

### **Screenshots**

N/A

### **Regression Notes**
There will be no impact on the screens, as AccountSettingsViewModel is not yet used by AccountSettingsFragment.

- [X] I have completed the Regression Notes.
- [X] I have considered adding accessibility improvements for my changes.
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

SowmiaS committed:
Creates AccountSettingsViewModel and AccountSettingsUiState

SowmiaS committed:
Adds AccountSettingsUiState StateFlow
  - Default value of UiState is updated with account information in AccountSettingsRepository

SowmiaS committed:
Adds empty methods for all ui events
  - on change of primary site, email, web address, password change.
  - on confirmation of user name change in server

SowmiaS committed:
Implements onAccountSettingsChange method

SowmiaS committed:
Implements onPrimarySiteChanged

SowmiaS committed:
Implements onEmailChanged

SowmiaS committed:
Implements onWebAddressChanged

SowmiaS committed:
Implements onPasswordChanged

SowmiaS committed:
Implements onUserNameConfirmed

SowmiaS committed:
Implements cancelPendingEmailChange

SowmiaS committed:
Fetch new settings on init

SowmiaS committed:
Creates AccountSettingsViewModelTest

SowmiaS committed:
Injects AccountSettingsViewModel

SowmiaS committed:
Adds unit test for AccountsSettingsViewModel

SowmiaS committed:
Adds methods to retrive siteNames, siteIds, homeURLOrHostNames

SowmiaS committed:
Merge branch 'issue/12372-refactor_account_settings_create_accountsettingsrepository' into issue/12372-refactor_account_settings_create_accountsettingsviewmodel

SowmiaS committed:
Makes changes to support suspending function getSitesAccessedViaWPComRest

SowmiaS committed:
Cancels fetch new settings job on updating account settings

SowmiaS committed:
Merge branch 'issue/12372-refactor_account_settings_create_accountsettingsrepository' into issue/12372-refactor_account_settings_create_accountsettingsviewmodel

SowmiaS committed:
Calls accountRepository cleanup method on viewmodel's onCleared

SowmiaS committed:
Merge branch 'issue/12372-refactor_account_settings_create_accountsettingsrepository' into issue/12372-refactor_account_settings_create_accountsettingsviewmodel

SowmiaS committed:
Fixes an issue where homeURLOrHostNames in primary site state returned sitenames instead of URLs

SowmiaS committed:
Fixes an issue where primarysiteviewmodel is not updated once the sites are retrived

SowmiaS committed:
Fixes detekt issues

SowmiaS committed:
Merge trunk into issue/12372-refactor_account_settings_create_accountsettingsviewmodel

SowmiaS committed:
Merge trunk into issue/12372-refactor_account_settings_create_accountsettingsviewmodel

SowmiaS committed:
Replaces Repository class with usecases

SowmiaS committed:
Moves unregister PushAccountSettingsUseCase object from eventbus to onCleared method

SowmiaS committed:
Adds onToastShown function to AccountSettingsViewModel

SowmiaS committed:
Creates AccountSettingsOptimisticUpdateHandler

SowmiaS committed:
Adds Inject constructor and replaces first with firstOrNull function

SowmiaS committed:
Adds back null safety to ui states

SowmiaS committed:
Injects AccountSettingsViewModel

SowmiaS committed:
fixes detekt and checkstyle issues in usecase

SowmiaS committed:
Removes checkstyle issues in AccountSettingsViewModelTest

SowmiaS committed:
Fixes checkstyle issues in AccountSettingsViewModel

SowmiaS committed:
Updates the uistate based on recently changed value

SowmiaS committed:
Fixes failing test - show change password progress dialog
- When the user tries to update the new password, show changing password progress dialog until it receives the server response

SowmiaS committed:
Replaces AccountRepository with usecases in AccountSettingsViewModelTest

SowmiaS committed:
Removes unneccessary update of uistate
- updates ui state only if optimistic update is applicable for that account settings change

SowmiaS committed:
Refactors AccountSettingsViewModelTest
   - Renames mUiState to uiState
   - Refactors test name to a meaningful sentence
   - Adds assertion to verify the email verification message
   - Remove redundant tests which check the initial UiState

SowmiaS committed:
Makes fetchNewSettingsJob private and fixes detekt issues

SowmiaS committed:
Reverts removal of binding method for ConversationNotificationsViewModel

SowmiaS committed:
Refactors AcountSettingsOptimisticUpdateHandler
  -  Renames class
  -  Adds comment to AcountSettingsOptimisticUpdateHandler
  -  Replaces with Kotlin forEach
  -  Replaces put() with indexing

SowmiaS committed:
Refactors AccountSettingsViewModel and adds default error message
  -  Renames getSitesAccessedViaWPComRest to getSites
  -  Reuses local variable account inside the function
  -  Adds default error message as Couldnt save your account settings

SowmiaS committed:
Merge branch 'trunk' into issue/12372-refactor_account_settings_create_accountsettingsviewmodel

SowmiaS commented:
@ashiagr This PR is now ready for review. Thanks 🙂

ashiagr commented:
Thanks, @SowmiaS!  It's on my radar. 😄 Will get to it asap.

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsViewModel.kt:
```diff
@@ -0,0 +1,275 @@
+        val siteIds
+            get() = sites?.map { it.siteId.toString() }?.toTypedArray()
+
+        val homeURLOrHostNames
```
Np: Unused property `homeURLOrHostNames` can be removed.

ashiagr replied:
Let's keep it assuming it'll be needed by the fragment.

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsViewModel.kt:
```diff
@@ -0,0 +1,275 @@
+    private fun getAccountSettingsUiState(): AccountSettingsUiState {
+        val account = getAccountUseCase.account
+        val siteViewModels = _accountSettingsUiState?.value?.primarySiteSettingsUiState?.sites
+        val showChangePasswordProgressDialog = _accountSettingsUiState?.value?.changePasswordSettingsUiState
```
Np: Unnecessary safe call can be removed on a non-null receiver of type `MutableStateFlow<AccountSettingsViewModel.AccountSettingsUiState>`

ashiagr replied:
Fixed in 62993d9

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsViewModel.kt:
```diff
@@ -0,0 +1,275 @@
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import org.wordpress.android.R.string
```
Np: Let's use `R.string.xxx` and remove string import here.

ashiagr replied:
Removed in ec32c89

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsViewModel.kt:
```diff
@@ -0,0 +1,275 @@
+                    )
+            )
+        }
+    }
```
Since we're updating primary site settings state, we might want to convey that purpose and rename this function to `updatePrimarySiteSettingsUiState()`.

ashiagr replied:
Updated in 1e95fa6

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsViewModel.kt:
```diff
@@ -0,0 +1,275 @@
+    }
+
+    fun onToastShown(toastMessage: String) {
+        if (_accountSettingsUiState.value.error.equals(toastMessage)) {
+            updateErrorUiState(null)
+        }
```
It's interesting to see how [transient messages are shown using UI state updates](https://developer.android.com/topic/architecture/ui-layer/events#consuming-trigger-updates). 👍 This pattern is not used in the codebase till now for showing the snackbars. I'll want to see it in action when we use the view model in the fragment. 🙌

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsOptimisticUpdateHandler.kt:
```diff
@@ -0,0 +1,77 @@
+        if (preferenceValue.size == 1) {
+            optimisticallyChangedPreferenceMap.remove(preferenceKey)
+        } else {
+            optimisticallyChangedPreferenceMap[preferenceKey] = preferenceValue.drop(1)
```
Since `preferenceValue` can default to an empty list, `preferenceValue.drop(1)` might result in `Requested element count $n is less than zero.` error. Can we check if the list is empty, the function should simply return with no action?

ashiagr replied:
Done in d581ca8

ashiagr commented on WordPress/src/test/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsViewModelTest.kt:
```diff
@@ -0,0 +1,570 @@
+                getAccountUseCase,
+                getSitesUseCase,
+                optimisticUpdateHandler
+        )
```
Can we replace it with `initialisViewModel()` defined at the end of the class?

ashiagr replied:
Replaced in e929f56

ashiagr commented on WordPress/src/test/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsViewModelTest.kt:
```diff
@@ -0,0 +1,570 @@
+                        getAccountUseCase,
+                        getSitesUseCase,
+                        optimisticUpdateHandler
+                )
```
`initialiseViewModel()` can be updated to include `usernameCanBeChanged` param as follows:

```
private fun initialiseViewModel(
   usernameCanBeChanged: Boolean = false
){
        whenever(getAccountUseCase.account.usernameCanBeChanged).thenReturn(usernameCanBeChanged)
        viewModel = AccountSettingsViewModel(
                resourceProvider,
                networkUtilsWrapper,
                TEST_DISPATCHER,
                fetchAccountSettingsUseCase,
                pushAccountSettingsUseCase,
                getAccountUseCase,
                getSitesUseCase,
                optimisticUpdateHandler
        )
    }
```

Then we can call `initialiseViewModel(usernameCanBeChanged = true)` to reduce number of lines in tests. This can be applied to all tests where we're initialising the view model.

ashiagr replied:
62993d9 reduces number of lines in tests.

ashiagr commented on WordPress/src/test/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsViewModelTest.kt:
```diff
@@ -0,0 +1,570 @@
+                    .withFailMessage("The initial primary site Id should be updated with the cached account information")
+                    .isEqualTo(getAccountUseCase.account.primarySiteId)
+        } ?: Assertions.fail("The UiState should not be empty")
+        Assertions.assertThat(uiState.userNameSettingsUiState.userName)
```
Let's import `assertThat` and `fail` and remove the `Assertions` quailfier as there aren't too many imports from the class.

ashiagr replied:
Done in 49a2f4b

ashiagr commented on WordPress/src/test/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsViewModelTest.kt:
```diff
@@ -0,0 +1,570 @@
+
+    // Web Address
+    @Test
+    fun `When the user tries to update a new webaddress, optimistically show the user as if the new web address is changed even before updating in the server`() =
```
Test descriptions are too long at places, they'll fail with ktlint warnings. 

This one can be simplified to:
`When a new web address is entered, then new web address is updated optimistically`
 
Related to this, we might want to split a test if we're testing multiple conditions

Eg. This test case 
`When the user has changed the username through a different screen and navigated back with new username, the new username should be updated and notified with the snackbar message`

can be split to test two conditions: 

`new username should be updated `
and 
`notified with the snackbar message`

`When username change confirmed from the server, then new username is updated`
`When username changed confirmed from the server, then username changed snackbar is shown`

ashiagr replied:
Done in 62993d9

SowmiaS replied:
@ashiagr Thanks for the detailed feedback 🙇

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/AccountSettingsViewModel.kt:
```diff
@@ -0,0 +1,275 @@
+import javax.inject.Named
+
+const val ONE_SITE = 1
+class AccountSettingsViewModel @Inject constructor(
```
Looks like the classes are not formatted, can you please format them?

ashiagr replied:
Formatted in ad9702f

ashiagr reviewed (commented):
👋  @SowmiaS 

Thank you for your hard work on adding the view model class in order to convert the Accounts Settings screen to MVVM! 

Changes look good overall. I added a few comments, some of them are generic and applicable in multiple places. I'll give it another pass once these are addressed.

I'll also trigger CI shortly so that you can fix lint issues.

SowmiaS committed:
Removes import of R.string and uses directly in the code

SowmiaS committed:
Renames getSites() to updatePrimarySiteSettingsUiState()

SowmiaS committed:
Replaces redundant AccountSettingsViewModel initialization code with initialisViewModel()

SowmiaS committed:
Imports assertThat and fail and removes the Assertions quailfier

SowmiaS committed:
Reformat the code
   - AccountSettingsViewModel
   - AccountSettingsViewModelTest

SowmiaS committed:
Refactors AccountSettingsViewModelTest
  - Renames to short test names
  - Splits the tests into multiple tests
  - Reduced no of lines inside the test function

SowmiaS committed:
Adds isEmpty guard check for removeFirstChange
  - removes optional Unit for update and removeFirstChange

SowmiaS committed:
Removes failMessage for the test
   - When the username change is confirmed from the server, then the snackbar is shown

SowmiaS committed:
Updates getAccountSettingsUiState function with the parameter isInitial and removes lint issue
 - unnecessary safe call

ashiagr commented:
Thank you once again for all the changes, @SowmiaS!  💪 
I reran CI and see a few checkstyle issues. Can you please take a look? We should be ready to land this PR once CI passes. 🎉

SowmiaS committed:
Fixes checkstyle

SowmiaS commented:
@ashiagr I was working on the checkstyle issues and updating the comments and marking the PR as Ready for Review.
Thanks for updating the comments. 🙇 
Sorry for the delay.
This PR is ready now. can you please rerun the CI?

ashiagr reviewed (approved):
Thanks, @SowmiaS! I should've waited for the re-review request. 😅 
With the CI passing, I'm merging this PR now.

AliSoftware commented:
Good news @SowmiaS: your fix has made its way to the version `20.1` that we started to progressively roll out to end users this morning 🎉  It should appear in the Play Store to all users between today and tomorrow.
